src/main/java/life/qbic/data/processing/registration/ProcessRegistrationRequest.java

```diff
@@ -40,53 +52,134 @@ public class ProcessRegistrationRequest extends Thread {
+    try (var content = Files.walk(workingTargetDir)) {
+      content.map(Path::toFile).sorted(Comparator.reverseOrder()).forEach(File::delete);
+    }
+  }
+
+  private static void processMeasurement(String measurementId,
+      Map<String, List<RegistrationMetadata>> aggregatedFilesByMeasurementId, Path workingTargetDir,
+      Path taskDir) throws IOException {
+    for (RegistrationMetadata metadataEntry : aggregatedFilesByMeasurementId.get(
+        measurementId)) {
+      Files.move(workingTargetDir.resolve(metadataEntry.file()),
+          taskDir.resolve(metadataEntry.file()));
+    }
+  }
+
+  private void moveBackToOrigin(Path target, Path usersHomePath, String reason) {
+    log.info("Moving back to original user directory: {}", usersHomePath);
+    try {
+      Path taskDir = createTaskDirectory();
+      Files.move(target, taskDir.resolve(target.getFileName()));
+      var errorFile = taskDir.resolve("error.txt").toFile();
+      errorFile.createNewFile();
+      Files.writeString(errorFile.toPath(), reason);
+      usersHomePath.resolve(usersErrorDirectory).toFile().mkdir();
+      Files.move(taskDir,
+          usersHomePath.resolve(usersErrorDirectory)
+              .resolve(taskDir.toFile().getName()));
+    } catch (IOException e) {
+      log.error("Cannot move task to user intervention: %s".formatted(
+          usersHomePath.resolve(usersErrorDirectory)), e);
+    }
+  }
+
+  private void validateFileEntries(Collection<RegistrationMetadata> metadata, Path request)
+      throws ValidationException {
+    for (RegistrationMetadata metadataEntry : metadata) {
+      if (!request.resolve(Paths.get(metadataEntry.file())).toFile().exists()) {
+        throw new ValidationException(
+            "Unknown file reference in metadata: %s".formatted(metadataEntry.file()),
+            ErrorCode.FILE_NOT_FOUND);
+      }
+    }
+    // To save the user from any trouble, let's also check if all dataset files are described by a metadata entry
+    var filesInMetadata = metadata.stream().map(RegistrationMetadata::file).toList();
+    var physicalFiles = request.toFile().listFiles();
+    for (File file : physicalFiles) {
+      // we ignore hidden files
+      if (file.isHidden()) {
+        continue;
+      }
+      if (file.getName().endsWith(metadataFileName)) {
+        continue;
+      }
+      if (!filesInMetadata.contains(file.getName())) {
+        throw new ValidationException(
+            "Found more files than described in the metadata file: %s".formatted(file.getName()),
+            ErrorCode.MISSING_FILE_ENTRY);
       }
     }
   }
 
+  private List<RegistrationMetadata> findAndParseMetadata(Path request) throws ValidationException {
+    Optional<File> metadataFile = findMetadataFile(request);
+    if (metadataFile.isEmpty()) {
+      throw new ValidationException("Metadata file does not exist",
+          ErrorCode.METADATA_FILE_NOT_FOUND);
+    }
```

Comment:
Do we check for empty lines between filenames and whitespace within the metadatafile?

Reply:
Very good point, I need to make sure empty lines are ignored